[Rails 6.0] Upgrade to rails 6.0

luisramos0 opened the pull request:
#### What? Why?

This PR upgrades OFN to rails 6. I think we can review this and just go forward when the 3 blocking PRs are merged in master..

Closes #7357

#### What should we test?
We need to make a thourough manual testing process.
Looking at the code changes, I'd test:
- permalinks
- stripe
- select2 boxes in the admin
- api tokens and app login/logout

This is dev-test PR re db2og: we need to test db backups.

#### Release notes
<!-- Write a one liner description of the change to be included in the release notes.
Every PR is worth mentioning, because you did it for a reason. -->

<!-- Please select one for your PR and delete the other. -->
Changelog Category: Technical changes
Upgrade to rails 6.0.

Matt-Yorkley commented on spec/lib/open_food_network/i18n_config_spec.rb:
```diff
@@ -5,6 +5,12 @@
   describe I18nConfig do
+    before do
+      allow(ENV).to receive(:[]).with("DATABASE_CLEANER_ALLOW_PRODUCTION").and_return("false")
+      allow(ENV).to receive(:[]).with("DATABASE_CLEANER_ALLOW_REMOTE_DATABASE_URL").
```
This looks a bit weird to me. It seems like these should both be the default values anyway? https://github.com/DatabaseCleaner/database_cleaner#safeguards

If we want to set these configs, it feels cleaner to just do this in `spec_helper.rb`:
```ruby
DatabaseCleaner.allow_production = true
DatabaseCleaner.allow_remote_database_url = true
```

luisramos0 replied:
this is actually coming from here: https://github.com/openfoodfoundation/openfoodnetwork/pull/7081/files
I just added it here to get the green build.

Matt-Yorkley replied:
Can we try dropping those lines that stub those two ENV vars and seeing what happens? I don't understand why they would be needed, as the defaults seem to be `false` for both of those options... :thinking:

Matt-Yorkley replied:
And it only breaks in this one spec file and not any others..?

luisramos0 replied:
yes, only breaks here. I will add to spec_helper and see. I'll do that in the 7081 PR first

Matt-Yorkley replied:
Ah! We're also stubbing ENV elsewhere in this file. I guess it's messing with the return values when `DatabaseCleaner` tries to check `ENV[:DATABASE_CLEANER_ALLOW_PRODUCTION]` as a boolean, for example.

Matt-Yorkley replied:
Alright. Curiosity: satisfied.

In that case, I'm not sure it makes sense to explicitly set those values to their defaults in `spec_helper.rb`? The bug is a consequence of the way we're stubbing of `ENV`, and not by the actual test suite configs being incorrect...

Matt-Yorkley replied:
We could leave it as it is, but maybe add an explanatory comment for why it's needed in this specific file.

luisramos0 replied:
ah, nice, thanks, makes sense! I did that in the PR https://github.com/openfoodfoundation/openfoodnetwork/pull/7081/files
It will come to this branch when rebased/merged from master :+1:

Matt-Yorkley commented on Gemfile.lock:
```diff
@@ -5,14 +5,14 @@ GIT
-  remote: https://github.com/openfoodfoundation/db2fog.git
-  revision: 8ceef362c64e6573d62d26db5ebb0c0f33cd3d61
-  branch: rails-5
+  remote: https://github.com/luisramos0/db2fog.git
```
I made you an admin on `openfoodfoundation/db2fog` @luisramos0.

We should move this branch to the repo :ok_hand:

Matt-Yorkley replied:
We'll probably want to `dev-test` this before deploying, backups are not covered in the suite.

luisramos0 replied:
done, thanks!
I have added the backups test to the description and added the dev-test label to this PR.

Matt-Yorkley commented on Gemfile.lock:
```diff
@@ -47,47 +47,59 @@ PATH
       actionpack (>= 4.0.0)
-    actionview (5.2.6)
-      activesupport (= 5.2.6)
+    actiontext (6.0.3.6)
```
We've got some new toys to play with here :smile:

Matt-Yorkley replied:
Yes, those are direct Rails dependencies in terms of the Gemfile, but we're now selectively loading them when booting the app :+1:

luisramos0 replied:
awesome :ok_hand:

Matt-Yorkley commented on Gemfile.lock:
```diff
@@ -404,18 +426,20 @@ GEM
-      activesupport (= 5.2.6)
+    rails (6.0.3.6)
+      actioncable (= 6.0.3.6)
+      actionmailbox (= 6.0.3.6)
```
The inclusion of this module in Rails core is a point of some contention in the Rails community :sweat_smile: 

It's only useful if you plan on running an online email service (`*cough* hey.com *cough*`), which 99% of Rails apps do not. It's quite large and I think some of it is enabled by default. We should have a quick look if there's any configs we need to add to **disable it** (I think there might be).

Matt-Yorkley replied:
Yeah ok, we need to update `application.rb` and remove `require 'rails/all'` and replace it with a list of require statements for all the modules _except_ `ActionMailbox`...

Otherwise we'll be loading a lot of pointless code, and it even **eats extra memory** (doing a job we don't need) :see_no_evil:

Matt-Yorkley replied:
I made a PR here: https://github.com/luisramos0/openfoodnetwork/pull/11

luisramos0 replied:
yes, nice, thanks :+1:

Matt-Yorkley commented on Gemfile.lock:
```diff
@@ -180,7 +193,7 @@ GEM
     compass-import-once (1.0.5)
       sass (>= 3.2, < 3.5)
-    compass-rails (4.0.0)
+    compass-rails (3.1.0)
```
It seems a bit weird that we have to forceably downgrade `compass-rails` here for Rails 6...? :thinking:

I guess it's not that important, as we'll be switching to Webpacker soon, and I think we'll be dropping these sass gems completely at that point.

luisramos0 replied:
nice, thanks for asking! I went back to it, removed the constraint in the Gemfile and typed  bundle update compass-rails. It works with v4, I must have left it in my initial attempts at getting bundle install to work with rails6. We are back to v4 :+1:

Matt-Yorkley replied:
Nice, yeah I had to do similar things getting bundler to install on previous upgrades :ok_hand:

Matt-Yorkley commented on spec/lib/stripe/account_connector_spec.rb:
```diff
@@ -15,7 +15,7 @@ module Stripe
 
       before do
-        allow(Stripe).to receive(:api_key) { "sk_test_12345" }
+        Stripe.api_key = "sk_test_12345"
```
Could we just set all these things once in `spec_helper.rb` or `environments/test.rb`, and drop all this code? Eg:
```ruby
Stripe.api_key = "sk_test_12345"
Stripe.client_id = "ca_abc123"
Stripe.publishable_key = "some_token"
```

Does that work? Is it preferable?

andrewpbrett replied:
Good thinking. Let's find out :) 4e7adb7d4c28092029f4874eed026cf6f8b791c4

luisramos0 replied:
For some reason I lost this commit, I had to replay it manually :-) Done, will see the build now.

luisramos0 replied:
ok, as far as I can see, it doesnt work, we just get all test broken again with "No API key defined"

luisramos0 replied:
for reference, the commit is here: https://github.com/luisramos0/openfoodnetwork/commit/dfcf311149a2e5bc54d8a274b03ccdce479834c1

andrewpbrett replied:
I had another quick try at this and didn't come to an immediate solution. IMO this is more of a nice-to-have and shouldn't hold up the whole PR from going to Test Ready - the lack of DRYness already existed and is unrelated to Rails 6. What do you think @Matt-Yorkley?

Matt-Yorkley commented on Gemfile:
```diff
@@ -109,6 +109,9 @@ gem 'jquery-migrate-rails'
 gem 'jquery-ui-rails', '~> 4.2'
 gem 'select2-rails', '~> 3.4.7'
+# Thor v0.20 works with both select2-rails 3.4.7 and railties 6.0.3.6
+#   To upgrade to thor v1 we need to upgrade select2-rails to v3.5 which requires some work
```
I think we should aim to completely ditch `select2`in the short-medium term and replace it with something more modern/lightweight/portable that doesn't depend directly on `jQuery`, like [SlimSelect](https://slimselectjs.com/).

It's not a job for today though...

luisramos0 replied:
I added this point to this issue here: https://github.com/openfoodfoundation/openfoodnetwork/issues/6510

luisramos0 commented:
I think I messed up this branch. I am rebasing from master now and will force push. Please refresh your local copy before adding commits :+1:

luisramos0 commented on Gemfile.lock:
```diff
@@ -6,13 +6,13 @@ GIT
-  revision: 8ceef362c64e6573d62d26db5ebb0c0f33cd3d61
-  branch: rails-5
+  revision: 5b63343847452f52aa42f7fc169d6ab3af57cda3
+  branch: rails-6
```
thanks @Matt-Yorkley for the permissions, done :+1:

luisramos0 commented:
ok, branch is rebased from master and healthy. The build is green. Matt's comments were addressed.
Only one point open re Stripe.api_key.

luisramos0 commented:
Hello, looks like Matt is ok with the Stripe.api_key point above, so this is ready for Review.

andrewpbrett reviewed (approved):
Technically it now has two approving reviews, but I think either @sauloperez or @mkllnk should give it a look before it goes to Test Ready

mkllnk commented on spec/support/request/stripe_stubs.rb:
```diff
@@ -37,7 +37,6 @@ def stub_payment_method_attach_request
 
   def stub_retrieve_payment_method_request(payment_method_id = "pm_1234")
     stub_request(:get, "https://api.stripe.com/v1/payment_methods/#{payment_method_id}")
-      .with(headers: { 'Authorization' => 'Bearer sk_test_12345' })
```
It's helpful to mention why you changed something instead what you changed. I can see the change in the code but may need some context why it's done or what the goal is.

luisramos0 replied:
This is great feedback Maikel. Thanks. I always add the "what" in the commit, the why is often, like in this case, very important!
To be honest i dont exactly understand why in some cases the authorization header is not necessary in the mock... but it's probably related to the api_key being set or not in the test case.

mkllnk reviewed (approved):
Looking good. :+1:

luisramos0 commented:
We can move to Test ready :+1: but meanwhile someone needs to do the dev-test (db backups).

luisramos0 commented:
rebased to resolve conflicts.

sauloperez commented on config/application.rb:
```diff
@@ -3,14 +3,14 @@
   "action_view/railtie",
   "action_mailer/railtie",
   "active_job/railtie",
   #"action_cable/engine", # Enable this when installing StimulusReflex
```
is this the only thing we'll skip compared to a `require 'rails/all'`?

sauloperez commented on config/application.rb:
```diff
@@ -1,6 +1,25 @@
+    require railtie
+  rescue LoadError
+  end
+end
```
I'm quite afraid of the unwanted side effects this could have in future upgrades. It's very easy to forget to add necessary requires and waste hours trying to find out why things are broken.

luisramos0 replied:
I see your point. I'd keep require rails/all for simplicity even if there's more code loaded.
But I am also ok with this solution.
We need a decision.

Matt-Yorkley replied:
ActionMailbox eats roughly 25MB of memory if it's running. Which doesn't sound like much until you realise that's _per worker_, so for UK (5 workers) that would be ~125MB of memory wasted for no reason. We definitely don't want that on all instances, right? Which means one way or another we need to avoid using `rails/all`.

It's now common practice to whitelist the Rails modules you're actually using (since ActionMailbox was added, hahaha) and having a single place where that list is explicitly declared seems pretty reasonable.

filipefurtad0 commented:
Hey @luisramos0!

The build appears to be broken:

```
rspec ./spec/features/admin/bulk_order_management_spec.rb:319 # 
  As an Administrator
  I want to be able to manage orders in bulk
 using page controls using drop down seletors distributor filter displays all line items when 'All' is selected from distributor filter

```

I thought it could be a flaky one, but ran it a couple of times - always failing... Staging anyway for now.

luisramos0 commented:
ok, I have fixed the spec, for some reason the helper is not available in the view after the rebase.

I removed "require rails/all" from spree/core.rb (rails gets required in application.rb and that's enough) which fixes the load issue in staging #7676.

I think this is good to be tested now.

luisramos0 commented:
rebased to resolve conflicts.

mkllnk commented:
https://staging.openfoodnetwork.org.au/api/v0/status/job_queue reports that Delayed Job is not working.

luisramos0 commented:
Thanks for the heads up Maikel, I am not sure what's the problem. Some problem with whenever and the schedule file maybe?

I wonder what's the impact on testing, some emails maybe not be going out, right?

filipefurtad0 commented:
Alright! Delayed_job is back:

![image](https://user-images.githubusercontent.com/49817236/120160219-ea76db80-c1ed-11eb-993c-f3c674ed4a34.png)

Emails are back up as well. Back to testing now :tada:

filipefurtad0 commented:
This might need another rebase, to account a change earlier today - #7711. Thanks :+1:
Let's see if we can push this upgrade through the next days...

luisramos0 commented:
oh, delayedjob is gone, nice.

Rebase done :+1: ready again.

filipefurtad0 commented:
Alright @luisramos0 ! I think we're good here:

On the cases mentioned above:

**Permalinks**  - OK
- Created products with same name -> permalinks differ and are unique
- Imported products and inventory with the same name -> permalinks are differ and are unique
- Manually edited permalinks for a given product -> changes are consistent through the shopfront

**Stripe**  - OK
- Setting up stripe
- Shopfront orders
- Backoffice orders
- Subscriptions
Manual payment authorization/authentication

**select2 boxes in the admin**  - OK
- Checked behavior/content for select2 boxes throughout the app

**api tokens and app login/logout**  - OK
- Checked used api tokens to retrieve content with:
`curl -X GET "https://staging.openfoodnetwork.org.uk/api/v0/products/bulk_products.json{20}?token=1030e27995f98b07c1ea0acea4914ee42ceb276f191410ec" -H "accept: application/json"
`
- Repeated after generating new tokens; verified the previous token is no longer valid

Full testing notes:
https://docs.google.com/document/d/1j5k4HZcrhO7d81Z9gq2_2nrX4PwyOGTN1mcSfHN8C4s/edit#heading=h.4ctypvfojpgm

I'm not entirely sure we're done with the dev-tests? Otherwise this can be moved to ready to go :tada:

luisramos0 commented:
Amazing. Thanks Filipe. I think it still needs a dev to test the backups are still working.

Matt-Yorkley commented:
Cool, so we just need a staging server where S3 backups are set up and actually working... I think that'll be UK :+1:

Matt-Yorkley commented:
AWS backups tested and working :heavy_check_mark::heavy_check_mark: 

Unless there's anything else we need to check i think this might be ready to go :astonished: :tada:

luisramos0 commented:
awesome, thank you!!!!

Matt-Yorkley commented:
Thanks @luisramos0!

sauloperez commented:
![](https://media.giphy.com/media/bSVbpBv6MiwuBT3Q6e/giphy.gif)
